### ISSUE TITLE
ArgoCD w/ SOPS leveraging AWS KMS

### ISSUE THREAD

**Issue by jmclean-starburst** [jkroepke/helm-secrets#165]
### Problem Statement

I am unable to leverage this plugin in its current state while using AWS KMS Keys and SOPS. do you know if there is support for this or am I just leveraging `helm-secrets` incorrectly

helm secrets version = v3.9.1

### Environment

- Helm Version:
- OS:
- Shell:

**Comment by jkroepke** [comment]
Hi,

in background, helm-secrets is a wrapper around sops. helm-secrets supports all features that are supported by sops.

Checkout https://github.com/mozilla/sops#kms-aws-profiles for integrate sops and AWS KMS. Additionally you need to provide credentials that the sops is authorized to call the AWS KMS keys.

In case you are running EKS, I'm recommend using [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html). You can also provide the environment variable AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY that holds credentials for a valid AWS IAM account.

**Comment by jmclean-starburst** [comment]
I had assumed as much....just things are a bit wonky w/ My ArgoCD referencing a sops encrypted file; one possibility is that I am forced to use the umbrella chart pattern because ArgoCD doesnt support values on a local filesystem/repo; only values within the repository containing the helm chart definition. The umbrella chart pattern allows values to be set on a local filesys/repo

More info https://argo-cd.readthedocs.io/en/stable/user-guide/helm/
```
Values files must be in the same git repository as the Helm chart. The files can be in a different location in which case it can be accessed using a relative path relative to the root directory of the Helm chart.
```

I have been hoping for this bad boy to close soon https://github.com/argoproj/argo-cd/pull/6280

**Comment by jkroepke** [comment]
Checkout https://github.com/jkroepke/helm-secrets/blob/main/docs/ARGOCD.md#known-limitations for some workarounds

**Comment by jkroepke** [comment]
In case you have a sops encrypted file with AWS KMS, you can use something like this.

```
secrets://https://raw.githubusercontent.com/org/repo/ref/pathtofile.yml
```


The `my-ns-w-secret/secret-name#secret-key` syntax is used in case you are using gpg or age encryption instead AWS KMS.

**Comment by jmclean-starburst** [comment]
@jkroepke thanks for your response! In this particular scenario, i am trying to access a private repository, and it seems that ArgoCD does not like having an environment variable in there; so the below doesnt work bc it takes `GITHUB_TOKEN` as a literal
```
secrets://https://$GITHUB_TOKEN@raw.githubusercontent.com/org/repo/ref/pathtofile.yml
```

Error:
```
Error: failed to fetch https://$GITHUB_TOKEN@raw.githubusercontent.com/org/repo/ref/pathtofile.yml: 404 Not Found
```

**Comment by jmclean-starburst** [comment]
more debug (just on my own PC - removing ArgoCD from the equation). The below works, if i remove the `secrets://` prefix, however the content added is an ENC blob (expected)
```
helm template mychart ./mychart -f secrets://https://$GITHUB_TOKEN@raw.githubusercontent.com/org/repo/ref/pathtofile.yml
error loading config: no matching creation rules found
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x221e632]

goroutine 1 [running]:
bytes.(*Buffer).Bytes(...)
	/usr/local/go/src/bytes/buffer.go:54
helm.sh/helm/v3/pkg/cli/values.readFile(0xc0002c6370, 0xa6, 0xc0002683c0, 0x3, 0x3, 0x3a9f108, 0x0, 0x1, 0xc000401940, 0xc0001298e0)
	/home/circleci/helm.sh/helm/pkg/cli/values/options.go:120 +0x162
helm.sh/helm/v3/pkg/cli/values.(*Options).MergeValues(0xc0002ddc20, 0xc0002683c0, 0x3, 0x3, 0xc00039e400, 0x2, 0x2)
	/home/circleci/helm.sh/helm/pkg/cli/values/options.go:48 +0x114
main.runInstall(0xc000536800, 0x2, 0x4, 0xc00027f540, 0xc0002ddc20, 0x28a5440, 0xc00000e018, 0xc0005f3a80, 0x11, 0x3618b98)
	/home/circleci/helm.sh/helm/cmd/helm/install.go:192 +0x331
main.newTemplateCmd.func2(0xc0000a2b00, 0xc000536800, 0x2, 0x4, 0x0, 0x0)
	/home/circleci/helm.sh/helm/cmd/helm/template.go:70 +0x125
github.com/spf13/cobra.(*Command).execute(0xc0000a2b00, 0xc0005367c0, 0x4, 0x4, 0xc0000a2b00, 0xc0005367c0)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000ceb00, 0xc00069c478, 0x1, 0x1)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/home/circleci/helm.sh/helm/cmd/helm/helm.go:93 +0x258
```

Note: I have `3.10.0` of the plugin installed locally
```
> helm plugin list | grep secrets
secrets	3.10.0 	This plugin provides secrets values encryption for Helm charts secure storing
```

**Comment by jmclean-starburst** [comment]
so...this is going to sound weird...but this works if I clone down your repository and add the plugin via `helm plugin install ./helm-secrets`, but if I install the plugin via the remote git repo (`helm plugin install https://github.com/jkroepke/helm-secrets`), i get something like the below:

```
Error: failed to parse secrets://secrets.yaml: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array into Go value of type map[string]interface {}
```

**Comment by jkroepke** [comment]
I have no idea here, except checkout if the plugin.yaml is equal at the plugin root

**Comment by jmclean-starburst** [comment]
FWIW - I found my issue (plaguing my soul for 3 days)

I referenced a helm-secrets version of `v3.9.1` in my ArgoCD Dockerfile, whereas it should have been `3.9.1`. painful...however I do think through this journey, I will try to contribute so we can use a k8s secret for fetching remote repos, instead of injecting directly within the URL (similar to GPG). more to come!

**Comment by jmclean-starburst** [comment]
https://github.com/jkroepke/helm-secrets/pull/168

**Comment by jkroepke** [comment]
Mention that the age support is available since 3.10.0

**Comment by Kiruthikameena** [comment]
@jmclean-starburst I am trying to use AWS KMS as well and facing some issues, could you please guide if something is wrong with the setup?

The issue is described here https://github.com/jkroepke/helm-secrets/issues/394. Thanks in advance!
